### PR TITLE
Initial impl of less-copy matbin parser

### PR DESCRIPTION
**Pull request by vswarte**

### CODE REVIEW

File: format/src/matbin/mod.rs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+#[allow(unused)]
+// Defines a material for instancing in FLVERs and such.
+// It does so by pointing at a shader and specifying the parameter/sampler
+// setup.
+pub struct Matbin<'buffer> {
+    bytes: &'buffer [u8],
+
+    pub header: &'buffer Header,
+    
+    pub parameters: &'buffer [Parameter],
```
Comment (vswarte) [inline format/src/matbin/mod.rs:37]: Don't really need to expose `pararameters` and `samplers` I guess

File: format/src/matbin/mod.rs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    ValueZerocopy,
+}
+
+#[derive(FromZeroes, FromBytes)]
+#[repr(C)]
+#[allow(unused)]
+// Defines a material for instancing in FLVERs and such.
+// It does so by pointing at a shader and specifying the parameter/sampler
+// setup.
+pub struct Matbin<'buffer> {
```
Comment (garyttierney) [inline format/src/matbin/mod.rs:32]: I don't think we need to name lifetimes when there's only one present and its bounds are obvious (e.g. everything is tied to the lifetime of the buffer).

File: format/src/matbin/mod.rs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    name_hash: U32<LE>,
+
+    /// Type of the value pointed at by value_offset
+    value_type: U32<LE>,
+
+    _padding18: Padding<16>,
+}
+
+#[derive(FromZeroes, FromBytes, Debug)]
+#[repr(C)]
```
Comment (garyttierney) [inline format/src/matbin/mod.rs:253]: `#[repr(packed)]` on `FromBytes` implementations.

File: format/src/matbin/mod.rs [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+    IO(#[from] io::Error),
+
+    #[error("Could not read string")]
+    String(#[from] ReadUtf16StringError),
+
+    #[error("Got unknown parameter type {0}")]
+    UnknownParameterType(u32),
+
+    #[error("Could not create reference to value")]
+    ValueZerocopy,
```
Comment (garyttierney) [inline format/src/matbin/mod.rs:23]: The only time this error can be produced (if bounds are already checked) is when the data is unaligned.

File: format/src/matbin/mod.rs [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+}
+
+#[derive(FromZeroes, FromBytes)]
```
Comment (garyttierney) [inline format/src/matbin/mod.rs:26]: Remove `FromBytes` and `FromZeroes` on this impl, it's not a type we can grab from a slice of bytes. Also no need for it to be `repr(C)`.

File: format/src/_matbin.rs [hunk 6]
```diff
@@ -1,12 +1,14 @@
...
+use thiserror::Error;
 
 use crate::io_ext::ReadFormatsExt;
 
-#[derive(Debug)]
+#[derive(Debug, Error)]
 pub enum MatbinError {
-    IO(io::Error),
+    #[error("Could not copy bytes {0}")]
+    IO(#[from] io::Error),
```
Comment (garyttierney) [inline format/src/_matbin.rs:11]: `Io`

File: format/src/matbin/mod.rs [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+            .field("shader_path", &self.shader_path())
+            .field("source_path", &self.source_path())
+            .field("header", self.header)
+            .field("parameters", &self.parameters)
+            .field("samplers", &self.samplers)
+            .finish()
+    }
+}
+
+#[derive(Debug)]
```
Comment (garyttierney) [inline format/src/matbin/mod.rs:148]: A custom impl of `Debug` would be better here, the `Debug` impl for F32<O> etc. isn't great.

**Review by garyttierney — CHANGES_REQUESTED** [review]
Some minor nits.

File: format/src/matbin/mod.rs (hunk 4 above)
Comment (vswarte) [reply format/src/matbin/mod.rs]: Renamed to `UnalignedValue` in f8ecf02d3d602cce65538df48def9046dd3255fc

File: format/src/matbin/mod.rs (hunk 5 above)
Comment (vswarte) [reply format/src/matbin/mod.rs]: 574b93d216ec8d05173f6aeb66908c934b48df2f removes it

File: format/src/_matbin.rs (hunk 6 above)
Comment (vswarte) [reply format/src/_matbin.rs]: Renamed in 5a8254ca826fda4d5c084f43e91449126f675289

File: format/src/matbin/mod.rs (hunk 2 above)
Comment (vswarte) [reply format/src/matbin/mod.rs]: Renamed per 808b7a96a9af4984a4ac229712b30be1cc015e48

File: format/src/matbin/mod.rs (hunk 1 above)
Comment (vswarte) [reply format/src/matbin/mod.rs]: Was fixed in rebase commit

File: format/src/matbin/mod.rs (hunk 3 above)
Comment (vswarte) [reply format/src/matbin/mod.rs]: 1fd01fd1a40e296400e33e7fe691ef86ef74c3d6 makes them all `repr(packed)`

**Review by garyttierney — APPROVED** [review]

**Pull request merged**